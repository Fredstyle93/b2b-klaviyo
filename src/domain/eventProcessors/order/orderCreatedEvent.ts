import { AbstractEvent } from '../abstractEvent';
import logger from '../../../utils/log';
import { Order, OrderCreatedMessage, OrderCustomerSetMessage, OrderState } from '@commercetools/platform-sdk';
import { getTypedMoneyAsNumber } from '../../../utils/get-typed-money-as-number';
import { getCustomerProfileFromOrder } from '../../../utils/get-customer-profile-from-order';
import { getOrderById } from '../../ctService';
import { mapAllowedProperties } from '../../../utils/property-mapper';
import config from 'config';

export class OrderCreatedEvent extends AbstractEvent {
    isEventValid(): boolean {
        const message = this.ctMessage as unknown as OrderCreatedMessage | OrderCustomerSetMessage;
        return (
            message.resource.typeId === 'order' &&
            this.isValidMessageType(message.type) &&
            this.hasExpectedMessageProperties(message)
        );
    }

    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as OrderCreatedMessage | OrderCustomerSetMessage;
        logger.info('Processing order created event');

        let order: Order;
        if ('order' in message) {
            order = message.order;
        } else {
            order = (await getOrderById(message.resource.id)) as Order;
        }

        const body: EventRequest = {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(order),
                    metric: {
                        name: config.get('order.metrics.placedOrder'),
                    },
                    value: this.context.currencyService.convert(
                        getTypedMoneyAsNumber(order?.totalPrice),
                        order.totalPrice.currencyCode,
                    ),
                    properties: mapAllowedProperties('order', { ...order }) as any,
                    unique_id: order.id,
                    time: order.createdAt,
                },
            },
        };

        const events: KlaviyoEvent[] = [{ body, type: 'event' }];

        this.getProductOrderedEventsFromOrder(events, order);

        return Promise.resolve(events);
    }

    private getCustomerProfile(order: Order): KlaviyoEventProfile {
        const profile: KlaviyoEventProfile = {};
        if (order.customerEmail) {
            profile.$email = order.customerEmail;
        }
        if (order.customerId) {
            profile.$id = order.customerId;
        }
        return profile;
    }

    private getProductOrderedEventsFromOrder(events: KlaviyoEvent[], order: Order) {
        order?.lineItems?.forEach((lineItem) => {
            events.push({
                body: {
                    data: {
                        type: 'event',
                        attributes: {
                            profile: this.getCustomerProfile(order),
                            metric: {
                                name: config.get('order.metrics.orderedProduct'),
                            },
                            value: getTypedMoneyAsNumber(lineItem.totalPrice),
                            properties: { ...lineItem },
                            unique_id: lineItem.id,
                            time: order.createdAt,
                        },
                    },
                },
                type: 'event',
            });
        });
    }

    private isValidState(orderState: OrderState): boolean {
        return Boolean(
            config.has('order.states.created') &&
                (config.get('order.states.created.placedOrder') as string[])?.includes(orderState),
        );
    }

    private isValidMessageType(type: string): boolean {
        return Boolean(
            config.has('order.messages.created') && (config.get('order.messages.created') as string[])?.includes(type),
        );
    }

    private hasExpectedMessageProperties(message: OrderCreatedMessage | OrderCustomerSetMessage) {
        return (
            !!(message as OrderCustomerSetMessage).customer ||
            (!!(message as OrderCreatedMessage).order &&
                (!!(message as OrderCreatedMessage).order?.customerEmail ||
                    !!(message as OrderCreatedMessage).order?.customerId) &&
                this.isValidState((message as OrderCreatedMessage).order?.orderState))
        );
    }
}
