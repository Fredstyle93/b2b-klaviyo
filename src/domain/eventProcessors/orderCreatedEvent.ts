import { AbstractEvent } from './abstractEvent';
import logger from '../../utils/log';
import { OrderCreatedMessage } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/message';
import { Order, OrderState } from '@commercetools/platform-sdk';
import { getTypedMoneyAsNumber } from '../../utils/get-typed-money-as-number';
import { getConfigProperty } from '../../utils/prop-mapper';
import { getCustomerProfileFromOrder } from '../../utils/get-customer-profile-from-order';

export class OrderCreatedEvent extends AbstractEvent {
    isEventValid(): boolean {
        const orderCreatedMessage = this.ctMessage as unknown as OrderCreatedMessage;
        return (
            orderCreatedMessage.resource.typeId === 'order' &&
            orderCreatedMessage.type === 'OrderCreated' &&
            !!orderCreatedMessage.order &&
            (!!orderCreatedMessage.order.customerEmail || !!orderCreatedMessage.order.customerId) &&
            this.isValidState(orderCreatedMessage.order?.orderState)
        );
    }

    generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const orderCreatedMessage = this.ctMessage as unknown as OrderCreatedMessage;
        logger.info('Processing order created event');

        const body = {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(orderCreatedMessage.order),
                    metric: {
                        name: this.getOrderMetric('OrderCreated'),
                    },
                    value: getTypedMoneyAsNumber(orderCreatedMessage.order?.totalPrice),
                    properties: { ...orderCreatedMessage.order } as any,
                    unique_id: orderCreatedMessage.order.id,
                    time: orderCreatedMessage.order.createdAt,
                },
            },
        };

        const events: KlaviyoEvent[] = [{ body, type: 'event' }];

        this.getProductOrderedEventsFromOrder(events, orderCreatedMessage.order);

        return Promise.resolve(events);
    }

    private getCustomerProfile(order: Order): KlaviyoEventProfile {
        const profile: KlaviyoEventProfile = {};
        if (order.customerEmail) {
            profile.$email = order.customerEmail;
        }
        if (order.customerId) {
            profile.$id = order.customerId;
        }
        return profile;
    }

    private isValidState(orderState: OrderState): boolean {
        return Boolean(getConfigProperty('order.createdStates', orderState));
    }

    private getProductOrderedEventsFromOrder(events: KlaviyoEvent[], order: Order) {
        order?.lineItems?.forEach((lineItem) => {
            events.push({
                body: {
                    data: {
                        type: 'event',
                        attributes: {
                            profile: this.getCustomerProfile(order),
                            metric: {
                                name: this.getOrderMetric('OrderedProduct'),
                            },
                            value: getTypedMoneyAsNumber(lineItem.totalPrice),
                            properties: { ...lineItem },
                            unique_id: lineItem.id,
                            time: order.createdAt,
                        },
                    },
                },
                type: 'event',
            });
        });
    }

    private getOrderMetric(type: string): string {
        return getConfigProperty('order.createdStates', type);
    }
}
