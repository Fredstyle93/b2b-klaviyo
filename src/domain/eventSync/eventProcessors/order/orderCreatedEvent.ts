import { AbstractEventProcessor } from '../abstractEventProcessor';
import logger from '../../../../utils/log';
import { Order, OrderCreatedMessage, OrderCustomerSetMessage, OrderState } from '@commercetools/platform-sdk';
import { getTypedMoneyAsNumber } from '../../../../utils/get-typed-money-as-number';
import { getCustomerProfileFromOrder } from '../../../../utils/get-customer-profile-from-order';
import { getOrderById } from '../../../../infrastructure/driven/commercetools/ctService';
import { mapAllowedProperties } from '../../../../utils/property-mapper';
import config from 'config';

export class OrderCreatedEvent extends AbstractEventProcessor {
    isEventValid(): boolean {
        const message = this.ctMessage as unknown as OrderCreatedMessage | OrderCustomerSetMessage;
        return (
            message.resource.typeId === 'order' &&
            this.isValidMessageType(message.type) &&
            this.hasExpectedMessageProperties(message)
        );
    }

    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as OrderCreatedMessage | OrderCustomerSetMessage;
        logger.info('Processing order created event');

        let order: Order;
        if ('order' in message) {
            order = message.order;
        } else {
            order = (await getOrderById(message.resource.id)) as Order;
        }

        const body: EventRequest = this.context.orderMapper.mapCtOrderToKlaviyoEvent(order, config.get('order.metrics.placedOrder'));

        const events: KlaviyoEvent[] = [{ body, type: 'event' }];

        this.getProductOrderedEventsFromOrder(events, order);

        return Promise.resolve(events);
    }

    private getProductOrderedEventsFromOrder(events: KlaviyoEvent[], order: Order) {
        order?.lineItems?.forEach((lineItem) => {
            events.push({
                body: this.context.orderMapper.mapOrderLineToProductOrderedEvent(lineItem, order),
                type: 'event',
            });
        });
    }

    private isValidState(orderState: OrderState): boolean {
        return Boolean(
            config.has('order.states.created') &&
                (config.get('order.states.created.placedOrder') as string[])?.includes(orderState),
        );
    }

    private isValidMessageType(type: string): boolean {
        return Boolean(
            config.has('order.messages.created') && (config.get('order.messages.created') as string[])?.includes(type),
        );
    }

    private hasExpectedMessageProperties(message: OrderCreatedMessage | OrderCustomerSetMessage) {
        return (
            !!(message as OrderCustomerSetMessage).customer ||
            (!!(message as OrderCreatedMessage).order &&
                (!!(message as OrderCreatedMessage).order?.customerEmail ||
                    !!(message as OrderCreatedMessage).order?.customerId) &&
                this.isValidState((message as OrderCreatedMessage).order?.orderState))
        );
    }
}
