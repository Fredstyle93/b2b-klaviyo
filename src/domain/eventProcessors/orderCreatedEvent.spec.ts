import { expect as exp } from 'chai';
import { mockDeep } from 'jest-mock-extended';
import { OrderCreatedEvent } from './orderCreatedEvent';
import { MessageDeliveryPayload } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/subscription';

describe('orderCreatedEvent > isEventValid', () => {
    it('should return valid when is an orderCreated message', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'order' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'OrderCreated' }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'order', { value: { customerId: '123-1230-123', orderState: 'Open' } }); //mock readonly property

        const event = OrderCreatedEvent.instance(ctMessageMock);

        exp(event.isEventValid()).to.be.true;
    });

    it.each`
        resource     | type              | order
        ${'invalid'} | ${'OrderCreated'} | ${{ customerId: '123-1230-123' }}
        ${'order'}   | ${'invalid'}      | ${{}}
        ${'order'}   | ${'OrderCreated'} | ${null}
    `('should return invalid when is not an orderCreated message', ({ resource, type, order }) => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: resource } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: type }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'order', { value: order }); //mock readonly property

        const event = OrderCreatedEvent.instance(ctMessageMock);

        exp(event.isEventValid()).to.be.false;
    });

    it('should return invalid when the orderState is invalid', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'order' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'OrderCreated' }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'order', {
            value: { customerId: '123-1230-123', orderState: 'FakeState' },
        }); //mock readonly property

        const event = OrderCreatedEvent.instance(ctMessageMock);

        exp(event.isEventValid()).to.be.false;
    });
});

describe('orderCreatedEvent > generateKlaviyoEvent', () => {
    it('should generate the klaviyo event for an order created message', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'order' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'OrderCreated' }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'order', {
            value: { customerId: '123-123-123', customerEmail: 'test@klaviyo.com', orderState: 'Open' },
        }); //mock readonly property

        const event = OrderCreatedEvent.instance(ctMessageMock);

        const klaviyoEvent = event.generateKlaviyoEvent();

        exp(klaviyoEvent).to.not.be.undefined;
        expect(klaviyoEvent.body).toMatchSnapshot();
    });
});
