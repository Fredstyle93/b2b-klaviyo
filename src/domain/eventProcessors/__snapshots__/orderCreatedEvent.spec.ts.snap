// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orderCreatedEvent > generateKlaviyoEvent should generate the klaviyo event for an order created message 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "Order created",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
      },
      "unique_id": undefined,
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`orderCreatedEvent > generateKlaviyoEvent should generate the klaviyo events for line items in an order created message 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "Order created",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "lineItems": [
          {
            "id": "123-123-123",
            "name": {
              "en": "Test product",
            },
            "totalPrice": {
              "centAmount": 1300,
              "currencyCode": "USD",
              "fractionDigits": 2,
              "type": "centPrecision",
            },
          },
        ],
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
      },
      "unique_id": undefined,
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`orderCreatedEvent > generateKlaviyoEvent should generate the klaviyo events for line items in an order created message 2`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "Ordered Product",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "id": "123-123-123",
        "name": {
          "en": "Test product",
        },
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
      },
      "unique_id": "123-123-123",
      "value": 13,
    },
    "type": "event",
  },
}
`;
